test/indexer/test_tdr.py

```diff
@@ -0,0 +1,211 @@
+        manifest, metadata = canned_result['manifest'], canned_result['metadata']
+        return TDRBundle(source=source,
+                         uuid=bundle_uuid,
+                         version=one(e['version'] for e in manifest if e['name'] == 'links.json'),
+                         manifest=manifest,
+                         metadata_files=metadata)
+
+    def _make_mock_tdr_tables(self,
+                              source: TDRSource,
+                              bundle_fqid: BundleFQID,
+                              *,
+                              tables: Mapping[str, JSONs] = None) -> None:
+        if tables is None:
+            tables = self._load_canned_file(bundle_fqid, 'tables.tdr')['tables']
+        for table_name, table_rows in tables.items():
+            self._make_mock_entity_table(source, table_name, table_rows['rows'])
+
+    def test_emulate_bundle(self):
+        with self.subTest('valid links'):
+            self._test_bundle(self.test_source)
+
+        links_content_column = self.tinyquery.tables_by_name[
+            f'{self.test_source.bq_name}.links'
+        ].columns['content'].values
+        links_content = json.loads(one(links_content_column))
+        link = first(link for link in links_content['links']
+                     if link['link_type'] == 'supplementary_file_link')
+
+        with self.subTest('invalid supp file entity id'):
+            link['entity']['entity_type'] = 'cell_suspension'
+            links_content_column[0] = json.dumps(links_content)
+            with self.assertRaises(RequirementError):
+                self._test_bundle(self.test_source, load_tables=False)
+
+        with self.subTest('invalid supp file entity type'):
+            link['entity']['entity_type'] = 'project'
+            link['entity']['entity_id'] += '_wrong'
+            links_content_column[0] = json.dumps(links_content)
+            with self.assertRaises(RequirementError):
+                self._test_bundle(self.test_source, load_tables=False)
+
+    def _test_bundle(self,
+                     source: TDRSource,
+                     *,
+                     load_tables: bool = True):
+        test_bundle = self._canned_bundle(source)
+        if load_tables:
+            self._make_mock_tdr_tables(source, test_bundle.fquid)
+        plugin = TestPlugin(source, self.tinyquery)
+        emulated_bundle = plugin.emulate_bundle(test_bundle.fquid)
+        self.assertEqual(test_bundle.fquid, emulated_bundle.fquid)
+        self.assertEqual(test_bundle.metadata_files.keys(),
+                         emulated_bundle.metadata_files.keys())
+
+        def key_prefix(key):
+            return key.rsplit('_', 1)[0]
+
+        for key, value in test_bundle.metadata_files.items():
+            # Ordering of entities is non-deterministic so "process_0.json" may
+            # in fact be "process_1.json", etc
+            self.assertEqual(1, len([k
```

Comment:
Newline after `[`.

IN commit `Wrap at 80 chars`

and check for other cases of this